### DISCUSSION TITLE
E/Z stereo chemistry not preserved when converting mol to smiles

### DISCUSSION THREAD

**Discussion by LiuCMU** [rdkit/rdkit#6807] [Q&A] [answered]
Hello, the SMILES does not reflect the E/Z configuration after I setting the configuration of the double bond. 

Here is the code for reproducing:
```
# starting with a molecule
smi4 = 'C1=CCC1'
mol4 = Chem.MolFromSmiles(smi4)
# the molecule and its indexes are visualized below
```
![image](https://github.com/rdkit/rdkit/assets/60156077/9e446050-4314-476c-ba43-7e4231085a7d)
![image](https://github.com/rdkit/rdkit/assets/60156077/3e03dc4e-ba6a-4a8e-b1ce-4bee390222f7)

```
# edit the molecule
# remove a bond
rwmol = Chem.RWMol(mol4)
b = rwmol.GetBondWithIdx(3)
b_type = b.GetBondType()
bond_atom1 = b.GetBeginAtomIdx()
bond_atom2 = b.GetEndAtomIdx()
rwmol.RemoveBond(bond_atom1, bond_atom2)

# add CH3 groups on each end
atom1_idx = rwmol.AddAtom(Chem.Atom('C'))
atom2_idx = rwmol.AddAtom(Chem.Atom('C'))
rwmol.AddBond(bond_atom1, atom1_idx, b_type)
rwmol.AddBond(bond_atom2, atom2_idx, b_type)

# get the mol object
mol42,  = Chem.GetMolFrags(rwmol, asMols=True)

# visualize the edited molecule and its indexes
```
![image](https://github.com/rdkit/rdkit/assets/60156077/f99c759f-d1a4-415e-beae-bd92712da757)
![image](https://github.com/rdkit/rdkit/assets/60156077/114f70ca-3719-4045-b53d-744f7b26d200)

```
# set the stereo Z configuration for the double bond
double_bond = mol42.GetBondBetweenAtoms(0, 1)
double_bond.SetStereo(BondStereo.STEREOZ)
double_bond.SetStereoAtoms(5,2)

print(mol42.GetBondWithIdx(0).GetStereo())  # output STEREOZ
# visualize the edited molecule with specified Z configuration
```
![image](https://github.com/rdkit/rdkit/assets/60156077/3fb6c3bd-48f3-4a89-af97-25e6f7c1cc86)

As verified by the molecul visualization and the double bond configuration, the double bond in this molecule should be in Z configuration. However, when converting the mol object to SMILES, there is no stereo information:
```
Chem.MolToSmiles(mol42, isomericSmiles=True)  # output CC=CCCC
```

You can see that when visualzing the molecle, the double bond is in Z configuration. However, converting mol object to SMILES does not preserve the stereo information. 

Side note: my issues is similar to #5637 and #4. I have been try to edit a ring molecule and reserve its double bond stereo information for about a week. Does anyone know how to tackle this? Any help would be much appreciated, thanks in advance!

**Comment by bp-kelley** [comment]
To complete the stereo, you need to indicate exactly what atoms are effected.  It easier to start with existing stereo chemistry from a smiles and see what the data model is:

```
>>> m = Chem.MolFromSmiles("C/C=C/C")
>>> m.Debug()
Atoms:
	0 6 C chg: 0  deg: 1 exp: 1 imp: 3 hyb: 4 arom?: 0 chi: 0
	1 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 0 chi: 0
	2 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 0 chi: 0
	3 6 C chg: 0  deg: 1 exp: 1 imp: 3 hyb: 4 arom?: 0 chi: 0
Bonds:
	0 0->1 order: 1 dir: 4 conj?: 0 aromatic?: 0
	1 1->2 order: 2 stereo: 3 stereoAts: (0 3) conj?: 0 aromatic?: 0
	2 2->3 order: 1 dir: 4 conj?: 0 aromatic?: 0
```

You will see that for bond #1 you have a double bond, with stereo==3 (StereoE) and that the stereo atoms indices  are (0,3)

So, how to you set these?

```
>>> m = Chem.MolFromSmiles("CC=CC")
>>> m.GetBondWithIdx(1).SetStereo(Chem.BondStereo.STEREOE)
>>> m.GetBondWithIdx(1).SetStereoAtoms(0,3)
>>> Chem.MolToSmiles(m)
'C/C=C/C'
```

**Comment by LiuCMU** [comment]
Hi @bp-kelley , thank you for the quick response!

Here is the debugging mode for the `mol42` object in the reproducing code, I got
```
>>>>  mol42.Debug()

Atoms:
	0 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 0 chi: 0
	1 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 0 chi: 0
	2 6 C chg: 0  deg: 2 exp: 2 imp: 2 hyb: 4 arom?: 0 chi: 0
	3 6 C chg: 0  deg: 2 exp: 2 imp: 2 hyb: 4 arom?: 0 chi: 0
	4 6 C chg: 0  deg: 1 exp: 1 imp: 3 hyb: 4 arom?: 0 chi: 0
	5 6 C chg: 0  deg: 1 exp: 1 imp: 3 hyb: 4 arom?: 0 chi: 0
Bonds:
	0 0->1 order: 2 stereo: 2 stereoAts: (5 2) conj?: 0 aromatic?: 0
	1 1->2 order: 1 conj?: 0 aromatic?: 0
	2 2->3 order: 1 conj?: 0 aromatic?: 0
	3 3->4 order: 1 conj?: 0 aromatic?: 0
	4 0->5 order: 1 conj?: 0 aromatic?: 0

>>>> Chem.MolToSmiles(mol42, isomericSmiles=True)
'CC=CCCC'
```

You see, although the bond with Idx 0 is a double bond with stereic configuration 2 (assuming Z), the output SMILES does not contain stereo configuration.

**Comment by LiuCMU** [comment]
Some additional context about this issue:

1. Here is my RDKit version information: `rdkit                     2023.03.3        py39h40f44b6_1    conda-forge`;
2. The stereo information only lost if my `mol42` object is obtained from an edited molecule object as in the reproducing code. If I initiate a mol object with a SMILES, then set the double bond stereo chemistry, the stereo information is reserved;
3. I used the following function for finding the indexes of atoms and bonds:
```
def visualize_mol_idx(mol):
    """Visualize a molecule as a graph with atom and bond indices"""
    # Create an empty graph
    G = nx.Graph()

    # Add nodes for each atom in the molecule
    for atom in mol.GetAtoms():
        atom_idx = atom.GetIdx()
        atom_label = atom.GetSymbol()
        G.add_node(atom_idx, label=f"{atom_label} ({atom_idx})")

    # Add edges for each bond in the molecule
    for bond in mol.GetBonds():
        bond_idx = bond.GetIdx()
        bond_type = str(bond.GetBondType())
        G.add_edge(bond.GetBeginAtomIdx(), bond.GetEndAtomIdx(), label=f"{bond_type} ({bond_idx})")

    # Draw the graph with labels for nodes and edges
    pos = nx.spring_layout(G)
    node_labels = nx.get_node_attributes(G, 'label')
    edge_labels = nx.get_edge_attributes(G, 'label')
    nx.draw(G, pos, with_labels=True, labels=node_labels, font_size=12, node_size=1000, node_color='white', edge_color='black')
    nx.draw_networkx_edge_labels(G, pos, edge_labels=edge_labels, font_size=10, label_pos=0.5)

    # Show the plot
    plt.show()
``` 

Please let me know if you have any sugguestions. Thank you very much!

**Comment by bp-kelley** [comment, marked as answer]
I admit to getting a little lost here, but the more complicated example needs to set a bond direction equivalent to the \ and / notations in smiles.   This appears to be due to the order of the bonds:

I've recreated your example given above:

```
m = Chem.RWMol()
m.AddAtom(Chem.Atom(6))
m.AddAtom(Chem.Atom(6))
m.AddAtom(Chem.Atom(6))
m.AddAtom(Chem.Atom(6))
m.AddAtom(Chem.Atom(6))
m.AddAtom(Chem.Atom(6))
m.AddBond(0,1, Chem.BondType.DOUBLE)
m.AddBond(1,2, Chem.BondType.SINGLE)
m.AddBond(2,3, Chem.BondType.SINGLE)
m.AddBond(4,5, Chem.BondType.SINGLE)
m.AddBond(0,5, Chem.BondType.SINGLE)
Chem.SanitizeMol(m)
Chem.MolToSmiles(m)
```


And now to set the cis/trans, note that this is a bit complicated due to the order of the atom in your molecule, but should get you on the right path:

```
m.GetBondBetweenAtoms(1,0).SetStereo(Chem.BondStereo.STEREOE)
m.GetBondBetweenAtoms(1,0).SetStereoAtoms(5,2)
# this wasn't required for the simpler example
m.GetBondBetweenAtoms(1,2).SetBondDir(Chem.BondDir.ENDUPRIGHT)
m.GetBondBetweenAtoms(0,5).SetBondDir(Chem.BondDir.ENDDOWNRIGHT)
Chem.MolToSmiles(m)
```

```
'CC/C=C/CC'
```

Now this seems pretty complicated to me.   We could probably work up what the bond directions should be for STEREOE and STEREOZ given a configuration of bonds, but I'm not sure this function currently exists in the RDKit.

**Reply by LiuCMU** [reply]
Thank you very much for the detailed information! I appreciate it. I was able to get SMILES with slash notation after setting bond directions. But I am not sure what does `ENDUPRIGHT` and `ENDDOWNRIGHT` represent? Can I understand that they force the two substituents to be on the same sides of a double bond? The best information I found about bond direction is at this link: https://www.rdkit.org/docs/cppapi/classRDKit_1_1Bond.html#a9279eea9c460523e13f0f4a9f5fc2dfe

For the same molecule, if created from SMILES, I can get the SMILES with stereo notation without setting bond directions. For example:
```
smi = 'CC=CCCC'
mol = Chem.MolFromSmiles(smi)

mol.GetBondWithIdx(1).SetStereo(BondStereo.STEREOZ)
mol.GetBondWithIdx(1).SetStereoAtoms(0, 3)
print(Chem.MolToSmiles(mol, isomericSmiles=True))  # output  C/C=C\CCC
```
Do you happen to know why is it not necessary to set bond directions in this case. When you say "this wasn't required for the simpler example", do you mean the complexity of molecule or the complexity of operations on the molecule object? Thank you!

**Reply by LiuCMU** [reply]
For whoever interested, `ENDUPRIGHT` and `ENDDOWNRIGHT` set the bonds to different sides of a double bond. Here is the detailed discussion of them: https://github.com/rdkit/rdkit/discussions/5176

**Comment by LiuCMU** [comment]
One more question arouse after setting the bond directions: the resulting SMILES represent different stereo configuration (cis/Z) when compared with the molecule double bond object (E). 

```
# construct a molecule
m = Chem.RWMol()
m.AddAtom(Chem.Atom(6))
m.AddAtom(Chem.Atom(6))
m.AddAtom(Chem.Atom(6))
m.AddAtom(Chem.Atom(6))
m.AddAtom(Chem.Atom(6))
m.AddAtom(Chem.Atom(6))
m.AddBond(0,1, Chem.BondType.DOUBLE)
m.AddBond(1,2, Chem.BondType.SINGLE)
m.AddBond(2,3, Chem.BondType.SINGLE)
m.AddBond(4,5, Chem.BondType.SINGLE)
m.AddBond(0,5, Chem.BondType.SINGLE)
Chem.SanitizeMol(m)

# set stereochemistry of the Double Bond
m.GetBondBetweenAtoms(1,0).SetStereo(Chem.BondStereo.STEREOE)
m.GetBondBetweenAtoms(1,0).SetStereoAtoms(5,2)
print(Chem.MolToSmiles(m))  # output: CCC=CCC

m.GetBondBetweenAtoms(1,2).SetBondDir(Chem.BondDir.ENDDOWNRIGHT)
m.GetBondBetweenAtoms(0,5).SetBondDir(Chem.BondDir.ENDDOWNRIGHT)
print(m.GetBondWithIdx(0).GetStereo())  # output: STEREOE
print(Chem.MolToSmiles(m))  # output: CC/C=C\CC, this SMILES actually represents a molecule with the cis/Z configuration
```
It feels like `Bond.SetStereo` and `Bond.SetBondDir` are independent. Calling the former method set the stereo of a bond but does not infuence the resulting SMILES. Calling the latter changes the resulting SMILES, but does not influence the stereo configuration of the bond in the molecule object. Ideally, should the bond directions be infered based on the bond stereo configuration?

I feel the issue is getting tricky. Anyone who knows this issue is very welcomed and much appreciated. Thanks!

**Comment by bp-kelley** [comment]
That was simply an error on my part, just because it "worked" didn't mean it was correct, you should always set the bond direction, that is the smiles model we use internally.  Apologies for the confusion.  You can see that it was set in my initial example, I just forgot to set them:

```
>>> m = Chem.MolFromSmiles("C/C=C/C")
>>> m.Debug()
Atoms:
	0 6 C chg: 0  deg: 1 exp: 1 imp: 3 hyb: 4 arom?: 0 chi: 0
	1 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 0 chi: 0
	2 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 0 chi: 0
	3 6 C chg: 0  deg: 1 exp: 1 imp: 3 hyb: 4 arom?: 0 chi: 0
Bonds:
	0 0->1 order: 1 dir: 4 conj?: 0 aromatic?: 0
	1 1->2 order: 2 stereo: 3 stereoAts: (0 3) conj?: 0 aromatic?: 0
	2 2->3 order: 1 dir: 4 conj?: 0 aromatic?: 0
```

See bonds 0 and 2

**Comment by LiuCMU** [comment]
Thank you for the prompt reply! I found a solution for setting the stereo configuration of a double bond based on our discussions. Though I still have a question about setting bond directions: why cannot the bond directions be infered based on the bond stereo configuration, but has to be set mannually afterwards?

The value information you provided is enough for me to move on, big thanks! @bp-kelley I will keep an eye on this in the future, too.
